File: spring/core/SpringBeanLifecycles/src/main/java/demos/lifecycle/hooks/via/interfaces/StockCheckEngineMock.java
package demos.lifecycle.hooks.via.interfaces;

import demos.lifecycle.common.StockCheckEngine;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.stereotype.Component;

@Component("stock")
public class StockCheckEngineMock implements StockCheckEngine, InitializingBean, DisposableBean {
    public int check(String itemNo) {
        return 200;
    }

    @Override
    public void destroy() throws Exception {
        System.out.println("Spring context just released a stock check engine");
    }

    @Override
    public void afterPropertiesSet() throws Exception {
        System.out.println("Spring context just created a stock check engine");
    }
}

File: spring/core/AnnotationDrivenDIWithKotlin/src/main/kotlin/demos/spring/notes/v10/StockCheckEngineMock.kt
package demos.spring.notes.v10

import demos.spring.notes.common.StockCheckEngine
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component("stock")
class StockCheckEngineMock : StockCheckEngine {
    private var testData: Map<String, Int> = emptyMap()
        @Value("#{dataSource.buildStockData()}")
        set(value) {
            field = value
        }

    override fun check(itemNo: String): Int {
        require(testData.containsKey(itemNo)) { "$itemNo does not exist in test map" }
        return testData[itemNo] ?: 0
    }
}
